containers/src/Data/IntMap/Strict/Internal.hs

```diff
@@ -1134,24 +1129,64 @@ fromAscListWithKey f (x0 : xs0) = fromDistinctAscList (combineEq x0 xs0)
 -- > fromDistinctAscList [(3,"b"), (5,"a")] == fromList [(3, "b"), (5, "a")]
 
 fromDistinctAscList :: [(Key,a)] -> IntMap a
-fromDistinctAscList []         = Nil
-fromDistinctAscList (z0 : zs0) = work z0 zs0 Nada
-  where
-    work (kx,!vx) []            stk = finish kx (Tip kx vx) stk
-    work (kx,!vx) (z@(kz,_):zs) stk = reduce z zs (branchMask kx kz) kx (Tip kx vx) stk
-
-    reduce :: (Key,a) -> [(Key,a)] -> Mask -> Prefix -> IntMap a -> Stack a -> IntMap a
-    reduce z zs _ px tx Nada = work z zs (Push px tx Nada)
-    reduce z zs m px tx stk@(Push py ty stk') =
-        let mxy = branchMask px py
-            pxy = mask px mxy
-        in  if shorter m mxy
-                 then reduce z zs m pxy (Bin pxy mxy ty tx) stk'
-                 else work z zs (Push px tx stk)
-
-    finish _  t  Nada = t
-    finish px tx (Push py ty stk) = finish p (link py ty px tx) stk
-        where m = branchMask px py
-              p = mask px m
+fromDistinctAscList = fromAscList
+    -- surprisingly, trying to exploit the lack of equal keys does not seem
+    -- to result in a speedup (with ghc-8.6.5 -O2)
 
 data Stack a = Push {-# UNPACK #-} !Prefix !(IntMap a) !(Stack a) | Nada
+
+-- | /O(n)/. Build a map from a list of key\/value pairs with monotonic keys
+-- and a combining function.
+--
+-- The precise conditions under which this function works are subtle:
+-- For any branch mask, keys with the same prefix w.r.t. the branch
+-- mask must occur consecutively in the list.
+
+fromMonoListWithKey :: (Key -> a -> a -> a) -> [(Key,a)] -> IntMap a
+fromMonoListWithKey f []               = Nil
+fromMonoListWithKey f ((kx,!vx) : zs1) = addAll' kx vx zs1
+  where
+    -- `addAll'` collects all keys equal to `kx` into a single value,
+    -- and then proceeds with `addAll`.
+    addAll' !kx vx [] = Tip kx vx
+    addAll' !kx vx ((ky,!vy) : zs)
+        | kx == ky
+        = let !v = f kx vy vx in addAll' ky v zs
+        -- inlined: | otherwise = addAll kx (Tip kx vx) (ky : zs)
+        | m <- branchMask kx ky
+        , Inserted ty zs' <- addMany' m ky vy zs
+        = addAll kx (linkWithMask m ky ty kx (Tip kx vx)) zs'
+
+    -- for `addAll` and `addMany`, kx is /a/ key inside the tree `tx`
+    -- `addAll` consumes the rest of the list, adding to the tree `tx`
+    addAll !kx !tx []
+        = tx
+    addAll !kx !tx ((ky,!vy) : zs)
+        | m <- branchMask kx ky
+        , Inserted ty zs' <- addMany' m ky vy zs
+        = addAll kx (linkWithMask m ky ty kx tx) zs'
+
+    -- `addMany'` is similar to `addAll'`, but proceeds with `addMany'`.
+    addMany' !m !kx vx [] = Inserted (Tip kx vx) []
+    addMany' !m !kx vx zs0@((ky,!vy) : zs)
```

Comment:
The bang on `vy` doesn't look right. If I remember correctly, we're not supposed to force values for duplicate keys; we force the result of combining them with the old values instead.

Reply:
Good point. I was working under the assumption that all values are supposed to be forced, but indeed we shouldn't make the function more strict than it currently is without a very good reason.

For your amusement (the results may look a bit surprising, but given that `fromList{,With}` works by inserting values into the map as it goes, the bottoms are unavoidable:
```
(using containers-0.6.2.1)
Prelude Data.IntMap.Strict> fromList [(1,undefined),(1,2)] `seq` ()
*** Exception: Prelude.undefined
Prelude Data.IntMap.Strict> fromAscList [(1,undefined),(1,2)] `seq` ()
()
Prelude Data.IntMap.Strict> fromListWith (\_ _ -> 2) [(1,undefined),(1,1)] `seq` ()
*** Exception: Prelude.undefined
Prelude Data.IntMap.Strict> fromListWith (\_ _ -> 2) [(1,1),(1,undefined)] `seq` ()
()
Prelude Data.IntMap.Strict> fromAscListWith (\_ _ -> 2) [(1,undefined),(1,1)] `seq` ()
()
Prelude Data.IntMap.Strict> fromAscListWith (\_ _ -> 2) [(1,1),(1,undefined)] `seq` ()
()
```

Reply:
done